### PR TITLE
Add scripts and structure

### PR DESCRIPTION
**Pull request by Marilyth**
Scripts and README.md for downloading and processing ERA5 data locally

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,45 @@
...
+pip -r install requirements.txt
+```
+4. Make sure cdo and nco are installed\
+   e.g. for Ubuntu/Debian
+```
+sudo apt-get install cdo nco netcdf-bin
+```
+
+# Usage
+1. After completing both setups, you can use this script as follows:
```
Comment (joernu76) [inline README.md:35]: I am not 100% sure, but in ReST the paragraph needs to be indented. It formats reasonably well as is with github, but the example is not indented properly.

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,10 @@
+*.tmp*
+*.nc*
+*.grib*
+*.save*
+*README I
+*Getting Started
```
Comment (joernu76) [inline .gitignore]: Are all these necessary?

File: bin/add_ancillary.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
+"""
+Copyright (C) 2012 by Forschungszentrum Juelich GmbH
```
Comment (joernu76) [inline bin/add_ancillary.py]: Please update!

File: bin/add_ancillary.py [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+    if len(arg) != 1:
+        print(oppa.get_usage())
+        exit(1)
+    if not os.path.exists(arg[0]):
+        print("Cannot find model data at", arg[1])
+        exit(1)
+    return opt, arg[0]
+
+
+def add_eqlat(ncin):
```
Comment (joernu76) [inline bin/add_ancillary.py]: please remove eqlat

File: bin/add_ancillary.py [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+        history = datetime.datetime.now().isoformat() + ":" + " ".join(sys.argv)
+        if hasattr(ncin, "history"):
+            history += "\n" + ncin.history
+        ncin.history = history
+        ncin.date_modified = datetime.datetime.now().isoformat()
+
+        if option.eqlat:
+            add_eqlat(ncin)
+
+        add_surface(ncin, "pressure", option.surface_pressure)
```
Comment (joernu76) [inline bin/add_ancillary.py]: Do we need the surface fields for the server?

File: bin/download_an_all.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+def ml():
+    c_ml.retrieve('reanalysis-era5-complete', {
+        'class': 'od',
+        'time': sys.argv[2],
+        'date': sys.argv[1],
+        'expver': '1',
+        'levelist': sorted(range(1, 138)),
+        'levtype': 'ml',
+        'param': [129.128, 130.128, 131.128, 132.128, 133.128, 135.128, 152.128, 155.128, 203.128, 246.128, 247.128, 248.128],
+        'stream': 'oper',
```
Comment (joernu76) [inline bin/download_an_all.py:28]: please factor out common parameters.

File: bin/download_an_all.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+c_ml = cdsapi.Client()
+c_pl = cdsapi.Client()
+c_pv = cdsapi.Client()
+c_tl = cdsapi.Client()
+c_sfc = cdsapi.Client()
+
+def ml():
+    c_ml.retrieve('reanalysis-era5-complete', {
+        'class': 'od',
+        'time': sys.argv[2],
```
Comment (joernu76) [inline bin/download_an_all.py]: please assign argv parameters to variables and use those later in the code, e.g.
```
date, time = sys.argv[1:3]
types = sys.argv[3:]
```

File: bin/download_an_all.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        'levtype': 'sfc',
+        'param': ["151.128", "165.128", "166.128", "186.128", "187.128", "188.128"],
+        'stream': 'oper',
+        'type': 'an',
+        "area": "0/0/-80/360",
+        "grid": "1.0/1.0",
+    }, f'grib/{sys.argv[1]}T{sys.argv[2]}.an.sfc.grib')
+
+threads = []
+if "ml" in sys.argv:
```
Comment (joernu76) [inline bin/download_an_all.py]: evtl.
```
for typ in ["ml", ...]:
    if typ in types:
        threads.append(Thread(target=globals()[typ]))
        threads[-1].start()
```

File: bin/get_data.sh [hunk 9]
```diff
@@ -0,0 +1,127 @@
+#!/bin/bash
+# Limit maximum threads to avoid potential issues
```
Comment (joernu76) [inline bin/get_data.sh]: ... to a reasonable number on large multi-core computers

File: bin/get_data.sh [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+then 
+    init_date="${DATE}T${TIME}"
+fi
+export time_units="hours since ${init_date}"
+
+# Download ml, sfc, pv and pt files
+echo "Downloading files, this might take a long time!"
+if [ ! -f grib/${BASE}.ml.grib ]; then
+    echo test
+    python bin/download_an_all.py $DATE $TIME ml &
```
Comment (joernu76) [inline bin/get_data.sh]: Das ist jetzt doppelt paralellisiert. Mach das doch bitte in python *oder* bash.

File: bin/get_data.sh [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+ncatted -O -a standard_name,height,o,c,atmosphere_altitude_coordinate $alfile
+ncap2 -s "height@units=\"km\";height=height/1000" $alfile $alfile-tmp
+mv $alfile-tmp $alfile
+ncks -7 -L 7 -C -O -x -v lev,sp,lnsp $alfile $alfile
+rm $tmpfile
+
+# model/surface levels
+ncks -O -d lev_2,0,0 -d lev_2,16,28,4 -d lev_2,32,124,2 $mlfile $tmpfile
+rm $mlfile
+nccopy -s -d7 $tmpfile $mlfile
```
Comment (joernu76) [inline bin/get_data.sh]: die konversion/kompression wird doch unten mit ncks en-passant erledigt, oder?

File: bin/get_data.sh [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+# convert grib to netCDF, set init time
+cdo -f nc4c copy grib/${BASE}.tl.grib $tlfile
+ncatted -a units,time,o,c,"${time_units}" $tlfile
+cdo -f nc4c copy grib/${BASE}.pv.grib $pvfile
+ncatted -a units,time,o,c,"${time_units}" $pvfile
+cdo -f nc4c copy ${GRIB} $mlfile
+ncatted -a units,time,o,c,"${time_units}" $mlfile
+
+# Add pressure and geopotential height to model levels file
+bin/add_pressure_gph.sh input=$mlfile pressure_units=Pa gph_units="m^2s^-2"
```
Comment (joernu76) [inline bin/get_data.sh:67]: Does this work in case "." is not in PATH ?

File: bin/get_data.sh [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+echo "Creating pressure level file..."
+cdo ml2pl,85000,50000,40000,30000,20000,15000,12000,10000,8000,6500,5000,4000,3000,2000,1000,500,100 $mlfile $plfile
+ncatted -O -a standard_name,plev,o,c,atmosphere_pressure_coordinate $plfile
+ncap2 -s "plev/=100;plev@units=\"hPa\"" $plfile $plfile-tmp
+mv $plfile-tmp $plfile
+ncks -7 -L 7 -C -O -x -v lev,sp,lnsp,nhyi,nhym,hyai,hyam,hybi,hybm $plfile $plfile
+
+echo "Creating potential temperature level file..."
+python bin/interpolate_missing_variables.py $mlfile $tlfile pt
+python bin/rename_standard.py $mlfile $tlfile
```
Comment (joernu76) [inline bin/get_data.sh:93]: Das rename_standard Skript ist seltsam benamt und mir ist nicht ganz klar warum man es braucht.
Am ml-file hast Du dich kaum rumgefummelt, warum sind die Attribute da drin besser?

File: bin/get_data.sh [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+ncatted -O -a standard_name,lev,o,c,atmosphere_potential_temperature_coordinate $tlfile
+ncatted -O -a standard_name,pv,o,c,ertel_potential_vorticity $tlfile
+ncks -O -7 -L 7 $tlfile $tlfile
+
+echo "Creating potential vorticity level file..."
+python bin/interpolate_missing_variables.py $mlfile $pvfile pv
+python bin/rename_standard.py $mlfile $pvfile
+ncatted -O -a standard_name,lev,o,c,atmosphere_ertel_potential_vorticity_coordinate $pvfile
+ncatted -O -a standard_name,pt,o,c,air_potential_temperature $plfile
+ncatted -O -a units,lev,o,c,"kelvin * meter ** 2 / kilogram / second" $pvfile
```
Comment (joernu76) [inline bin/get_data.sh]: Is this correct? we typically deal in microkelvin * ....
the values in MSS level drop-down should be 2/4 or so. The file I downloaded with your script has a level of "2000", but should be (probably?) 2e-6 in the units given here...

File: bin/download_an_all.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        "grid": "1.0/1.0",
+    }, f'grib/{sys.argv[1]}T{sys.argv[2]}.an.ml.grib')
+
+def pl():
+    c_pl.retrieve('reanalysis-era5-complete', {
+        'class': 'od',
+        'time': sys.argv[2],
+        'date': sys.argv[1],
+        'expver': '1',
+        'levelist': 'all',
```
Comment (joernu76) [inline bin/download_an_all.py]: all?

File: bin/download_an_all.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        'levtype': 'pl',
+        'param': 'all',
```
Comment (joernu76) [inline bin/download_an_all.py]: all?

**Review by joernu76 — CHANGES_REQUESTED** [review]

**Comment by ReimarBauer** [comment]
Which License can we add? Same as for MSS ?

File: bin/download_an_all.py (hunk 7 above)
Comment (ReimarBauer) [reply bin/download_an_all.py]: see argparse

File: bin/get_data.sh (hunk 12 above)
Comment (Marilyth) [reply bin/get_data.sh:67]: I am pretty sure it does. As it isn't in my path.

Comment (joernu76) [reply bin/get_data.sh:67]: Strange. Worked for me as well, but I do not know why...

File: bin/download_an_all.py (hunk 7 above)
Comment (joernu76) [reply bin/download_an_all.py]: True, but might be overkill here.

File: bin/add_ancillary.py (hunk 3 above)
Comment (Marilyth) [reply bin/add_ancillary.py]: To what though? I am an absolute idiot when it comes to copyrights.
Is this still a copyright by FZJ? Is it a copyright by MSS? Is it a copyright by you and me?

Comment (joernu76) [reply bin/add_ancillary.py]: Adopt year and authors. Maybe Reimar has additional comments. See his license comment. I'd handle this exactly as we did with MSS.

File: bin/get_data.sh (hunk 14 above)
Comment (joernu76) [reply bin/get_data.sh]: Using grib_dump, I get the following information, which is consistent to my understanding:

```
[...]
  # Potential vorticity surface 10-9 K m2 kg-1 s-1 (grib1/local/ecmf/3.table , grib1/3.table)  
  indicatorOfTypeOfLevel = 117;
  level = 2000;
[...]
```
Where do you got "your" units from? Maybe there really is a bug in the ECMWF documentation?

I would specify this as `mPVU`.

Comment (Marilyth) [reply bin/get_data.sh]: You are right, it is wrong.
I divided it by 1000 to convert it to PVU and corrected the units attribute to reflect that.

Comment (Marilyth) [reply bin/get_data.sh]: In general PV is not in PVU (microkelvin) but kelvin https://apps.ecmwf.int/codes/grib/param-db?id=60
Except for the levels which are also not in PVU but in nanokelvin. Generally this was confusing me.

File: bin/get_data.sh (hunk 13 above)
Comment (Marilyth) [reply bin/get_data.sh:93]: Einige Standardnamen und Units sind in ml festgelegt und in allen anderen nicht.
Statt 50 Zeilen lang alle Standardnamen und Units in allen files zu ersetzen, ersetze ich die 1 mal in ml und benutze das Skript nach der Interpolation.

**Review by joernu76 — APPROVED** [review]

**Pull request merged**